### PR TITLE
FirebaseBackingStore: improve performance for large backing stores

### LINKED ISSUE #2307: Pathological slow downs caused by backingStores
I believe this is a well known problem, but I wanted to have a tracking ticket. Please add detail as desired.

### PR DESCRIPTION
**Pull request by mykmartin**
Should fix #2307

This will break mutation notifications between disparate Collections ostensibly observing the same data. The backing store no longer receives update notifications from Firebase, because that would require pulling the full collection locally, avoiding which is the explicit intent of this new class. So if we have two arcs running on different devices with handles that reference the same entity, if one of them writes a new value to that id, the other will *not* receive a notification.

This doesn't seem to be a blocking issue in our current usage patterns.

### CODE REVIEW

**Comment by shans** [comment]
Thanks Mick! This break is not something I'd anticipated, but it makes sense given the limitations of Firebase.

Berni and I are going to talk through whether this is an acceptable compromise for performance in an hour or two. It might e.g. just mean we need to use pouch in multi-device situations for now. I'll update the PR with the outcome of that discussion.

**Comment by seefeldb** [comment]
But if one device adds another entity to a collection, the other gets notified?

File: src/runtime/storage/firebase-storage.ts [hunk 1]
```diff
@@ -1467,3 +1477,131 @@ class FirebaseBigCollection extends FirebaseStorageProvider implements BigCollec
...
+          return data;
+        }
+      }
+      // Returning an empty object deletes the Firebase node.
+      return {};
+    }, undefined, false);
+  }
+
+  async get(id: string) {
+    const snapshot = await this.childRef(id).once('value');
```
Comment (shans) [inline src/runtime/storage/firebase-storage.ts]: could you not .get() here to register for mutations?

Comment (mykmartin) [reply src/runtime/storage/firebase-storage.ts]: Yes, if you want to view mutations on a specific existing entity, we can attach a listener for that. For a referenceMode collection to observe mutations to any of its items, we'll need to attach a listener for each item in the collection. No idea if that has performance implications for the Firebase servers.

**Comment by mykmartin** [comment]
Berni: yes, changes to containment should still propagate because the set of references in a refMode collection is held in a separate Firebase table, to which the refMode collection still has a standard set-wide listener attached.

File: src/runtime/storage/firebase-storage.ts (hunk 1 above)
Comment (shans) [reply src/runtime/storage/firebase-storage.ts]: OK, got it. Performance considerations aside, would that fix the break you described? Or do I misunderstand the problem?

Comment (mykmartin) [reply src/runtime/storage/firebase-storage.ts]: Should do. I'll update the PR and run some tests to verify.

Comment (shans) [reply src/runtime/storage/firebase-storage.ts]: Thanks!

**Comment by shaper** [comment]
Are we intending to merge this?

**Comment by shans** [comment]
I discussed the potential breakages with Berni and Scott; they agreed that they're not sufficiently serious to prevent this going in.

For context: edits to entities won't propagate *across contexts* (i.e. they still should within the same arc); so edits need to be modeled as a remove/clear and an add in order to work properly.

This will probably only break some of our sharing demos, and breakages should be pretty easy to fix.

@seefeldb @sjmiles feel free to chime in if I'm misrepresenting you!

**Pull request merged**